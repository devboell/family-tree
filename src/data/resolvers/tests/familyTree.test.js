import knex from '../../connectors'
import FamilyTreeModel from '../../models/familyTree'
import {
  transformPartnerInfo,
  mergePartnerInfo,
  mergePartners,
  prepTable,
  familyTree,
} from '../familyTree'

describe('familyTree resolver', () => {
  it('transformPartnerInfo', () => {
    const fixture = { id: 1, name: 'John', partnerId: 2, relationId: 1 }
    const expected = { id: 1, name: 'John', partnerInfo: { partnerId: 2, relationId: 1 } }
    expect(transformPartnerInfo(fixture)).toEqual(expected)
  })

  it('mergePartnerInfo, null partner', () => {
    const acc = { partnerInfo: [] }
    const fixture = { id: 1, name: 'John', partnerInfo: { partnerId: null, relationId: null } }
    const expected = {
      id: 1,
      name: 'John',
      partnerInfo: [],
    }
    expect(mergePartnerInfo(acc, fixture)).toEqual(expected)
  })

  it('mergePartnerInfo, single partner', () => {
    const acc = { partnerInfo: [] }
    const fixture = { id: 1, name: 'John', partnerInfo: { partnerId: 2, relationId: 1 } }
    const expected = {
      id: 1,
      name: 'John',
      partnerInfo: [
        { partnerId: 2, relationId: 1 },
      ],
    }
    expect(mergePartnerInfo(acc, fixture)).toEqual(expected)
  })

  it('mergePartnerInfo, multiple partners', () => {
    const acc = { id: 1, name: 'John', partnerInfo: [{ partnerId: 2, relationId: 1 }] }
    const fixture = { id: 1, name: 'John', partnerInfo: { partnerId: 3, relationId: 2 } }
    const expected = {
      id: 1,
      name: 'John',
      partnerInfo: [
        { partnerId: 2, relationId: 1 },
        { partnerId: 3, relationId: 2 },
      ],
    }
    expect(mergePartnerInfo(acc, fixture)).toEqual(expected)
  })

  it('mergePartners', () => {
    const fixture = [
      { id: 1, name: 'John', partnerId: 2, relationId: 1 },
      { id: 1, name: 'John', partnerId: 3, relationId: 2 },
    ]
    const expected = {
      id: 1,
      name: 'John',
      partnerInfo: [
        { partnerId: 2, relationId: 1 },
        { partnerId: 3, relationId: 2 },
      ],
    }
    expect(mergePartners(fixture)).toEqual(expected)
  })
})

describe('familyTree resolver, async', () => {
  let familyTreeTable

  beforeEach(async () => {
    await knex.migrate.rollback()
    await knex.migrate.latest()
    await knex.seed.run()
    familyTreeTable = await FamilyTreeModel.familyTreeTable(1)
  })

  afterEach(async () => {
    await knex.migrate.rollback()
  })

  afterAll(() => {
    knex.destroy()
  })

  it('prepTable', () => {
    expect(prepTable(familyTreeTable)).toMatchSnapshot()
  })

  it('familyTree, id 1', async () => {
    const tree = await familyTree(undefined, { id: 1 })
    expect(tree).toMatchSnapshot()
  })

  it('familyTree, id 9', async () => {
    const tree = await familyTree(undefined, { id: 9 })
    expect(tree).toMatchSnapshot()
  })
})
