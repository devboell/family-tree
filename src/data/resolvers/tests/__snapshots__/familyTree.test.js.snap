// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`familyTree resolver, async familyTree, id 1 1`] = `
Object {
  "bornToId": null,
  "id": 1,
  "level": 0,
  "name": "William I",
  "partners": Array [
    Object {
      "children": Array [
        Object {
          "bornToId": 1,
          "id": 3,
          "level": 1,
          "name": "William Rufus",
          "partners": Array [],
        },
        Object {
          "bornToId": 1,
          "id": 4,
          "level": 1,
          "name": "Henry I",
          "partners": Array [
            Object {
              "children": Array [
                Object {
                  "bornToId": 2,
                  "id": 6,
                  "level": 2,
                  "name": "William Adelin",
                  "partners": Array [],
                },
                Object {
                  "bornToId": 2,
                  "id": 7,
                  "level": 2,
                  "name": "Empress Matilde",
                  "partners": Array [
                    Object {
                      "children": Array [
                        Object {
                          "bornToId": 3,
                          "id": 9,
                          "level": 3,
                          "name": "Henry II",
                          "partners": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "bornToId": 4,
                                  "id": 11,
                                  "level": 4,
                                  "name": "Richard I",
                                  "partners": Array [],
                                },
                                Object {
                                  "bornToId": 4,
                                  "id": 12,
                                  "level": 4,
                                  "name": "John I",
                                  "partners": Array [],
                                },
                              ],
                              "partner": Object {
                                "birthDate": 1122,
                                "bornToId": null,
                                "deathDate": 1204,
                                "gender": "F",
                                "id": 10,
                                "isMonarch": 0,
                                "name": "Eleanor of Aquitane",
                                "title": "Queen of England",
                              },
                            },
                          ],
                        },
                      ],
                      "partner": Object {
                        "birthDate": 1113,
                        "bornToId": null,
                        "deathDate": 1151,
                        "gender": "M",
                        "id": 8,
                        "isMonarch": 0,
                        "name": "Geoffrey Plantagenet",
                        "title": "Count of Anjou",
                      },
                    },
                  ],
                },
              ],
              "partner": Object {
                "birthDate": 1080,
                "bornToId": null,
                "deathDate": 1118,
                "gender": "F",
                "id": 5,
                "isMonarch": 0,
                "name": "Matilde of Scotland",
                "title": "Queen of England",
              },
            },
            Object {
              "children": Array [],
              "partner": Object {
                "birthDate": 1103,
                "bornToId": null,
                "deathDate": 1151,
                "gender": "F",
                "id": 14,
                "isMonarch": 0,
                "name": "Adeliza of Louvain",
                "title": "Queen of England",
              },
            },
          ],
        },
        Object {
          "bornToId": 1,
          "id": 13,
          "level": 1,
          "name": "Robert Curthose",
          "partners": Array [],
        },
        Object {
          "bornToId": 1,
          "id": 15,
          "level": 1,
          "name": "Adela of Normandy",
          "partners": Array [
            Object {
              "children": Array [
                Object {
                  "bornToId": 6,
                  "id": 17,
                  "level": 2,
                  "name": "Stephen of Blois",
                  "partners": Array [],
                },
              ],
              "partner": Object {
                "birthDate": 1045,
                "bornToId": null,
                "deathDate": 1102,
                "gender": "M",
                "id": 16,
                "isMonarch": 0,
                "name": "Stephen II Henry",
                "title": "Count of Blois",
              },
            },
          ],
        },
      ],
      "partner": Object {
        "birthDate": 1031,
        "bornToId": null,
        "deathDate": 1083,
        "gender": "F",
        "id": 2,
        "isMonarch": 0,
        "name": "Matilde of Flanders",
        "title": "Queen of England",
      },
    },
  ],
}
`;

exports[`familyTree resolver, async familyTree, id 9 1`] = `
Object {
  "bornToId": 3,
  "id": 9,
  "level": 0,
  "name": "Henry II",
  "partners": Array [
    Object {
      "children": Array [
        Object {
          "bornToId": 4,
          "id": 11,
          "level": 1,
          "name": "Richard I",
          "partners": Array [],
        },
        Object {
          "bornToId": 4,
          "id": 12,
          "level": 1,
          "name": "John I",
          "partners": Array [],
        },
      ],
      "partner": Object {
        "birthDate": 1122,
        "bornToId": null,
        "deathDate": 1204,
        "gender": "F",
        "id": 10,
        "isMonarch": 0,
        "name": "Eleanor of Aquitane",
        "title": "Queen of England",
      },
    },
  ],
}
`;

exports[`familyTree resolver, async prepTable 1`] = `
Array [
  Object {
    "bornToId": null,
    "id": 1,
    "level": 0,
    "name": "William I",
    "partnerInfo": Array [
      Object {
        "partnerId": 2,
        "relationId": 1,
      },
    ],
  },
  Object {
    "bornToId": 1,
    "id": 3,
    "level": 1,
    "name": "William Rufus",
    "partnerInfo": Array [],
  },
  Object {
    "bornToId": 1,
    "id": 4,
    "level": 1,
    "name": "Henry I",
    "partnerInfo": Array [
      Object {
        "partnerId": 5,
        "relationId": 2,
      },
      Object {
        "partnerId": 14,
        "relationId": 5,
      },
    ],
  },
  Object {
    "bornToId": 2,
    "id": 6,
    "level": 2,
    "name": "William Adelin",
    "partnerInfo": Array [],
  },
  Object {
    "bornToId": 2,
    "id": 7,
    "level": 2,
    "name": "Empress Matilde",
    "partnerInfo": Array [
      Object {
        "partnerId": 8,
        "relationId": 3,
      },
    ],
  },
  Object {
    "bornToId": 3,
    "id": 9,
    "level": 3,
    "name": "Henry II",
    "partnerInfo": Array [
      Object {
        "partnerId": 10,
        "relationId": 4,
      },
    ],
  },
  Object {
    "bornToId": 4,
    "id": 11,
    "level": 4,
    "name": "Richard I",
    "partnerInfo": Array [],
  },
  Object {
    "bornToId": 4,
    "id": 12,
    "level": 4,
    "name": "John I",
    "partnerInfo": Array [],
  },
  Object {
    "bornToId": 1,
    "id": 13,
    "level": 1,
    "name": "Robert Curthose",
    "partnerInfo": Array [],
  },
  Object {
    "bornToId": 1,
    "id": 15,
    "level": 1,
    "name": "Adela of Normandy",
    "partnerInfo": Array [
      Object {
        "partnerId": 16,
        "relationId": 6,
      },
    ],
  },
  Object {
    "bornToId": 6,
    "id": 17,
    "level": 2,
    "name": "Stephen of Blois",
    "partnerInfo": Array [],
  },
]
`;
